File: 2021/day6/solution.py
class Lanternfish:
    def __init__(self):
        self.counts = {0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0}
        self.new_fish_cycle = 8
        self.repeated_fish_cycle = 6

    def advance(self) -> None:
        new_counts = {0:0, 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0}
        for i in range(0, self.new_fish_cycle+1):
            if i == self.repeated_fish_cycle:
                new_counts[self.repeated_fish_cycle] = self.counts[0] + self.counts[i+1]
            elif i == self.new_fish_cycle:
                new_counts[self.new_fish_cycle] = self.counts[0]
            else:
                new_counts[i] = self.counts[i+1]

        self.counts = new_counts
    
    def total(self) -> int:
        total = 0
        for v in self.counts.values():
            total += v
        return total

def get_input() -> Lanternfish:
    lanternfish = Lanternfish()
    with open('input.txt', 'r') as file:
        for line in file:
            raw_nums = line.strip().split(',')
            for n in raw_nums:
                key = int(n)
                lanternfish.counts[key] += 1
    
    return lanternfish

def get_cycle_count(l: Lanternfish, cycles: int) -> int:
    for _ in range(cycles):
        l.advance()

    return l.total()

p1 = get_cycle_count(get_input(), 80)
print(p1)

p2 = get_cycle_count(get_input(), 256)
print(p2)

File: 2021/day6/solution.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <map>

using namespace std;

class LanternFish {
private:
    map<int, long long> ages;
public:
    explicit LanternFish(const string& s) {
        char delim = ',';
        string temp;
        for (char i : s) {
            if (i == delim) {
                int key = stoi(temp);
                if (ages.count(key) > 0) {
                    ages[key] += 1;
                } else {
                    ages[key] = 1;
                }
                temp = "";
            } else {
                temp.push_back(i);
            }
        }
        if (!temp.empty()) {
            int key = stoi(temp);
            if (ages.count(key) > 0) {
                ages[key] += 1;
            } else {
                ages[key] = 1;
            }
        }
    }

    void AdvanceDay() {
        map<int, long long> newAges = {};
        long long recycledParents = 0;

        for (auto i = 0; i <= 8; i++) {
            int nextIdx = i-1;
            if (nextIdx < 0) {
                nextIdx = 8;
            }

            if (ages.count(i) > 0) {
                newAges[nextIdx] = ages[i];
                if (nextIdx == 8) {
                    recycledParents = ages[i];
                }
            } else {
                newAges[nextIdx] = 0;
            }
        }

        if (recycledParents > 0) {
            if (newAges.count(6) > 0) {
                newAges[6] += recycledParents;
            } else {
                newAges[6] = recycledParents;
            }
        }

        ages = newAges;
    }

    long long Total() {
        long long result = 0;
        for (auto pair : ages) {
            result += pair.second;
        }
        return result;
    }
};

string getRawLanternFish(const string& filePath) {
    ifstream inputFile(filePath);
    vector<string> lines;
    if (!inputFile) {
        cerr << "Failed to open input file." << endl;
        return {};
    }
    string line;
    while (getline(inputFile, line)) {
        lines.push_back(line);
    }
    return lines[0];
}

long long getTotalLanternFishCount(LanternFish l, int days) {
    for (auto i = 0; i < days; i++) {
        l.AdvanceDay();
    }
    return l.Total();
}


int main() {
    cout << "Part 1: " + to_string(getTotalLanternFishCount( LanternFish(getRawLanternFish("input.txt")), 80)) << endl;
    cout << "Part 2: " + to_string(getTotalLanternFishCount( LanternFish(getRawLanternFish("input.txt")), 256)) << endl;
}
